### PR TITLE
[deep link] Fix an issue that selected linkdata is not highlighted in domain view and path view

### PR DESCRIPTION
**Pull request by hangyujin**
Fix this issue by saving linkDatas by domain and path, instead of calling a getter every time.

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read the [Flutter Style Guide] _recently_, and have followed its advice.
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or there is a reason for not adding tests.

![build.yaml badge]

If you need help, consider asking for help on [Discord].

[Contributor Guide]: https://github.com/flutter/devtools/blob/master/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[CLA]: https://cla.developers.google.com/
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[build.yaml badge]: https://github.com/flutter/devtools/actions/workflows/build.yaml/badge.svg

### CODE REVIEW

File: packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_controller.dart [hunk 1]
```diff
@@ -294,8 +296,16 @@ class DeepLinksController extends DisposableController {
   final selectedLink = ValueNotifier<LinkData?>(null);
   final pagePhase = ValueNotifier<PagePhase>(PagePhase.emptyState);
 
-  List<LinkData>? allValidatedLinkDatas;
-  final displayLinkDatasNotifier = ValueNotifier<List<LinkData>?>(null);
+  /// These are all link datas before applying displayOptions.
+  List<LinkData> allValidatedLinkDatas = <LinkData>[];
+  List<LinkData> allValidatedLinkDatasbyDomain = <LinkData>[];
+  List<LinkData> allValidatedLinkDatasbyPath = <LinkData>[];
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_controller.dart:302]: use var and remove type on LHS

File: packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_controller.dart [hunk 2]
```diff
@@ -150,9 +150,10 @@ class DeepLinksController extends DisposableController {
   String get applicationId =>
       _androidAppLinks[selectedVariantIndex.value]?.applicationId ?? '';
 
+  @visibleForTesting
   List<LinkData> get getLinkDatasByPath {
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_controller.dart:154]: not part of this PR, but we should avoid starting method names / getters with "get". Can you rename all methods / getters that violate that best practice?
https://dart.dev/effective-dart/design#avoid-starting-a-method-name-with-get

File: packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_controller.dart [hunk 3]
```diff
@@ -459,7 +470,17 @@ class DeepLinksController extends DisposableController {
...
+      applyFilters();
+    }
+  }
+
+  void applyFilters() {
+    displayLinkDatasNotifier.value = getFilterredLinks(allValidatedLinkDatas);
+    linkDataByDomainNotifier.value =
+        getFilterredLinks(allValidatedLinkDatasbyDomain);
+    linkDataByPathNotifier.value =
+        getFilterredLinks(allValidatedLinkDatasbyPath);
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_controller.dart:483]: since these three seem to always be updated in tandem, what about creating a dart object that contains all three, and then we can have a single value notifier instead of three? Or we could use a record with named elements if we don't want to create another dart class:

```
typedef ValidatedLinkData = ({List<LinkData> all, List<LinkData> byDomain, List<LinkData> byPath});
...
final validatedLinkData = (all: [], byDomain: [], byPath: []);
final filteredValidatedLinkData = ValueNotifier<ValidatedLinkData>((all: [], byDomain: [], byPath: []);
```

Comment (hangyujin) [reply packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_controller.dart]: done

File: packages/devtools_app/lib/src/screens/deep_link_validation/deep_link_list_view.dart [hunk 4]
```diff
@@ -366,23 +366,23 @@ class _AllDeepLinkDataTable extends StatelessWidget {
...
                   viewType: TableViewType.pathView,
-                  linkDatas: controller.getLinkDatasByPath,
+                  linkDatas: controller.displayLinkDatasNotifier.value.byPath,
                   controller: controller,
                 ),
                 _DataTable(
                   viewType: TableViewType.singleUrlView,
-                  linkDatas: linkDatas!,
+                  linkDatas: controller.displayLinkDatasNotifier.value.all,
                   controller: controller,
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/deep_link_validation/deep_link_list_view.dart:386]: use the `linkDatas` var

File: packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_controller.dart [hunk 5]
```diff
@@ -294,8 +297,15 @@ class DeepLinksController extends DisposableController {
...
 
-  List<LinkData>? allValidatedLinkDatas;
-  final displayLinkDatasNotifier = ValueNotifier<List<LinkData>?>(null);
+  /// These are all link datas before applying displayOptions.
+  var validatedLinkDatas =
+      ValidatedLinkDatas(all: [], byDomain: [], byPath: []);
+
+  /// These are link datas actually displayed in the data table after filtering by displayOptions.
+  final displayLinkDatasNotifier = ValueNotifier<ValidatedLinkDatas>(
+    ValidatedLinkDatas(all: [], byDomain: [], byPath: []),
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_controller.dart:306]: nit: can you use validatedLinkDatas as the starting value for this notifier

File: packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_controller.dart [hunk 6]
```diff
@@ -405,20 +415,23 @@ class DeepLinksController extends DisposableController {
...
+      byDomain: linkDatasByDomain(linkdata),
+      byPath: linkDatasByPath(linkdata),
+    );
     displayOptionsNotifier.value = displayOptionsNotifier.value.copyWith(
-      domainErrorCount: getLinkDatasByDomain
+      domainErrorCount: validatedLinkDatas.byDomain
           .where((element) => element.domainErrors.isNotEmpty)
           .length,
-      pathErrorCount: getLinkDatasByPath
+      pathErrorCount: validatedLinkDatas.byDomain
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_controller.dart:428]: this should be byPath

**Review by kenzieschmoll — APPROVED** [review]
